File: BinarySearchTree.py
class BinarySearchTree(object):
    def __init__(self, key):
        self.key = key
        self.left = None
        self.right = None

    def find(self, item):
        if item == self.key:
            return self
        elif item < self.key and self.left:
            return self.left.find(item)
        elif item > self.key and self.right:
            return self.right.find(item)
        else:
            return None

    def find_min(self):
        # 从根节点开始只要有左儿子就向左进行查找，终止点是最小元素
        if self.left:
            return self.left.find_min()
        else:
            return self

    def find_max(self):
        # 从根节点开始只要有右儿子就向右进行查找，终止点是最大元素，该函数没有用到递归
        tree = self
        if tree:
            while tree.right:
                tree = tree.right
        return tree

    def insert(self, item):
        # 先用 find 查找，如果找到，则什么都不做。如果没找到则将其插入遍历路径的最后一点
        if item < self.key:
            if self.left:
                self.left.insert(item)
            else:
                tree = BinarySearchTree(item)
                self.left = tree
        elif item > self.key:
            if self.right:
                self.right.insert(item)
            else:
                tree = BinarySearchTree(item)
                self.right = tree

    def delete(self, item):
        # 参考 http://www.cnblogs.com/linxiyue/p/3624597.html
        # 有三种情况
        # 第一种：节点是树叶，直接删除
        # 第二种：节点只有一个儿子，将该节点的 parent 指向节点的儿子
        # 第三种：节点有两个儿子，用其又子树的最小数据代替此节点的数据，并将其右节点的最小数据删除
        if self.find(item):
            if item < self.key:
                # 分情况进入不同递归
                self.left = self.left.delete(item)
                return self
            elif item > self.key:
                # 分情况进入不同递归
                self.right = self.right.delete(item)
                return self
            elif self.left and self.right:
                # 找到元素并且该元素有左右子节点，第三种情况
                key = self.right.find_min().key
                self.key = key
                self.right = self.right.delete(key)
                # 删除该元素节点右子树中最小元素并且把该元素节点的值设为右子树最小值
                return self
            else:
                # 找到元素且元素只有一个子树或者没有子树
                # 直接跳过该子树
                if self.left:
                    return self.left
                else:
                    return self.right
        else:
            # 未在树中找到该元素
            return self

File: Java/src/me/xlui/data_structure/tree/BinarySearchTree.java
package me.xlui.data_structure.tree;

@SuppressWarnings("unchecked")
public class BinarySearchTree<E extends Comparable> implements Tree<E> {
	private Node root;

	public BinarySearchTree() {
		this.insert((E) new Integer(30));
		this.insert((E) new Integer(15));
		this.insert((E) new Integer(41));
		this.insert((E) new Integer(33));
		this.insert((E) new Integer(50));
		/*
		 * 树结构：
		 *         30
		 *       15  41
		 *          33 50
		 */
	}

	public boolean hasElement(E element) {
		return this.find(this.root, element) != null;
	}

	public boolean hasElementIter(E element) {
		return this.findIter(this.root, element) != null;
	}

	public E findMax() throws Exception {
		Node node = this.findMax(this.root);
		if (node != null)
			return (E) node.data;
		else
			throw new Exception("BST is empty!");
	}

	public E findMin() throws Exception {
		Node node = this.findMin(this.root);
		if (node != null)
			return (E) node.data;
		else
			throw new Exception("BST is Empty!");
	}

	public void insert(E element) {
		this.root = this.insert(this.root, element);
	}

	public boolean delete(E element) {
		try {
			this.root = this.delete(this.root, element);
			return true;
		} catch (Exception e) {
			e.printStackTrace();
			return false;
		}
	}

	@Override
	public void preOrderTraversal() {
		this.preOrderTraversal(this.root);
	}

	@Override
	public void inOrderTraversal() {
		this.inOrderTraversal(this.root);
	}

	@Override
	public void postOrderTraversal() {
		this.postOrderTraversal(this.root);
	}

	@Override
	public void levelOrderTraversal() {

	}

	//**************************
	//   start private method
	//**************************

	/**
	 * 递归实现查找
	 */
	private Node find(Node node, E element) {
		if (node == null)
			return null;
		if (element.compareTo(node.data) > 0)
			return find(node.right, element);
		else if (element.compareTo(node.data) < 0)
			return find(node.left, element);
		else
			return node;
	}

	/**
	 * 迭代实现查找
	 */
	private Node findIter(Node node, E element) {
		while (node != null) {
			if (element.compareTo(node.data) > 0)
				node = node.right;
			else if (element.compareTo(node.data) < 0)
				node = node.left;
			else
				return node;
		}
		return null;
	}

	// 查找最大元素的迭代函数
	private Node findMax(Node node) {
		if (node != null)
			while (node.right != null)
				node = node.right;
		return node;
	}

	// 查找最小元素的递归函数
	private Node findMin(Node node) {
		if (node == null)
			return null;
		if (node.left == null)
			return node;
		return findMin(node.left);
	}

	// 插入，并保证二叉搜索树有序。
	private Node insert(Node node, E element) {
		if (node == null) {
			// 结点为空，创建并插入数据
			node = new Node();
			node.data = element;
		} else {
			if (element.compareTo(node.data) > 0)
				node.right = insert(node.right, element);
			else if (element.compareTo(node.data) < 0)
				node.left = insert(node.left, element);
		}
		return node;
	}

	/**
	 * 分三种情况：
	 * 1、删除的结点是叶子节点。将其父结点的指针为 null
	 * 2、删除的结点只有一个儿子。将其父结点的指针指向要删除结点的儿子结点
	 * 3、删除的结点有两个儿子。用右子树的 最小元素 或者左子树的 最大元素 替代要删除的结点
	 */
	private Node delete(Node node, E element) throws Exception {
		if (node == null) {
			throw new Exception("Not Found!");
		} else if (element.compareTo(node.data) < 0) {
			node.left = delete(node.left, element);
		} else if (element.compareTo(node.data) > 0) {
			node.right = delete(node.right, element);
		} else {
			if (node.left != null && node.right != null) {
				// 要删除的结点有两个儿子结点
				Node tmp = findMin(node.right);    // 找到右子树的最小结点
				node.data = tmp.data;
				node.right = delete(node.right, (E) node.data);
			} else {
				node = (node.left != null) ? node.left : node.right;
			}
			/*
			if (node.left != null) {
				// 要删除的结点有左儿子结点
				node = node.left;
			} else if (node.right != null) {
				// 要删除的结点有右儿子结点
				node = node.right;
			} else {
				// 要删除的结点没有儿子结点
				node = null;
			}
			*/
		}

		return node;
	}

	private void preOrderTraversal(Node node) {
		if (node != null) {
			System.out.print(node.data + " ");
			preOrderTraversal(node.left);
			preOrderTraversal(node.right);
		}
	}

	private void inOrderTraversal(Node node) {
		if (node != null) {
			inOrderTraversal(node.left);
			System.out.print(node.data + " ");
			inOrderTraversal(node.right);
		}
	}

	private void postOrderTraversal(Node node) {
		if (node != null) {
			postOrderTraversal(node.left);
			postOrderTraversal(node.right);
			System.out.print(node.data + " ");
		}
	}

	private static final class Node<E> {
		E data;
		Node left;
		Node right;
	}
}
